### PR TITLE
rocon_app_utilities

### PR DESCRIPTION
**Pull request by jihoonl**
Related issues #137 #139 #153 #138 
- rapp indexer
- rapp indexer tests

http://192.168.10.129/rocondoc_dev/rocon_app_utilities/html/index.html

Document is set as github page

http://robotics-in-concert.github.io/rocon_app_platform/

### CODE REVIEW

File: rocon_app_utilities/package.xml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0"?>
+<package>
+  <name>rocon_app_utilities</name>
+  <version>0.0.0</version>
+  <description>The rocon_app_utilities package</description>
+
+  <maintainer email="jihoonlee.in@gmail.com">Jihoon Lee</maintainer>
+  <author email="jihoonlee.in@gmail.com">Jihoon Lee</author>
```
Comment (stonier) [inline rocon_app_utilities/package.xml]: We are starting to follow the convention of writing authors without email. Maintainers should be contactable, not authors. Update?

File: rocon_app_utilities/src/rocon_app_utilities/indexer.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+          :raises: RappNotExistException: the given rapp name does not exist
+        '''
+        if not rapp_name in self.raw_data:
+            raise RappNotExistException(str(rapp_name) + ' does not exist')
+
+        rapp = self._resolve(rapp_name)
+
+        return rapp
+
+    def get_compatible_rapps(self, uri=rocon_uri.RoconURI()):
```
Comment (stonier) [inline rocon_app_utilities/src/rocon_app_utilities/indexer.py]: Just use the string argument as input. I bounced on this for a while and Jack pushed for strings at the same time as I started realising how much a headache it is having to deal with these objects all the time. Strings are much easier to fling around. Use the library for parsing, use it for validation - everything else is strings. 

Also most http manipulation is done via strings, only passing them to urllib or similar if some processing needs to be done on the string.

File: rocon_app_utilities/src/rocon_app_utilities/indexer.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        rapp.inherit(parent)
+        stack.append(parent.name)
+
+        return self._resolve_recursive(rapp, parent.get_parent(), stack)
+
+    def to_dot(self):
+        '''
+            returns the dot graph format. Not Implemented Yet.
+        '''
+        # TODO
```
Comment (stonier) [inline rocon_app_utilities/src/rocon_app_utilities/indexer.py:194]: Raising `NotImplentedError` is probably a good convention for us on stubs. See this [page](http://programmers.stackexchange.com/questions/231397/is-it-conventional-to-raise-a-notimplementederror-for-methods-whose-implementati) for a discussion.

File: rocon_app_utilities/src/rocon_app_utilities/rapp.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+        
+    t = str(impl + ' ' + ance)
+    return is_impl, is_ance, t 
+
+
+def validate_rapp_field(data):
+    '''
+        Validate each field. E.g) check rocon uri. Check the linked file exist 
+    '''
+    #  TODO
```
Comment (stonier) [inline rocon_app_utilities/src/rocon_app_utilities/rapp.py:115]: Again, NotImplemtedError

File: rocon_app_utilities/src/rocon_app_utilities/rapp.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+
+class Rapp(object):
+    '''
+        Rocon(or Robot) App definition. 
+    '''
+    __slots__ = ['name', 'data', 'type', 'is_impl', 'is_ance', 'ancestor_name']
+
+    def __init__(self, name, filename=None):
+        self.name = name
```
Comment (stonier) [inline rocon_app_utilities/src/rocon_app_utilities/rapp.py]: Is this the one you will shift to `ancestor_name`?

File: rocon_app_utilities/src/rocon_app_utilities/rapp.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+            self.load_from_file(filename)
+            #validate_rapp_field(self.dat)
+            self.classify()
+
+
+    def __str__(self):
+        return str(self.name + ' - ' + self.type)
+
+
+    def get_name(self):
```
Comment (stonier) [inline rocon_app_utilities/src/rocon_app_utilities/rapp.py]: I've finally given up my c++ tendencies to write a get_name on top of a name variable. It's alot of extra noise for something that isn't actually protecting anything anyway (no such thing as privates). 

There is a very strong convention for allowing people to access the variables directly. And if you want to replace the getter/setter mechanism more formally, use [python properties](http://stackoverflow.com/questions/6618002/python-property-versus-getters-and-setters). If you want to go one step further and make sure variables can't be accessed and manipulate many of them at once, you need to really dig into decorators (see rocon uri module).

In conclusion, let's drop the redundant get_ and set_ calls where not really useful.

File: rocon_app_utilities/src/rocon_app_utilities/rapp.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        '''
+          loads rapp data from file. and classifies itself.
+
+          :param filename: absolute path to rapp definition
+          :type filename: str
+        '''
+        self.data = load_yaml_from_file(filename)
+        self.classify()
+
+    def inherit(self, rapp):
```
Comment (stonier) [inline rocon_app_utilities/src/rocon_app_utilities/rapp.py:182]: Should this be internal? i.e. _inherit?

File: rocon_app_utilities/src/rocon_app_utilities/rapp.py [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+        del self.data['parent_name']
+
+        for attribute in INHERITABLE_ATTRIBUTES:
+            if not attribute in self.data and attribute in rapp.data: 
+                self.data[attribute] = rapp.data[attribute]
+
+        self.classify()
+
+
+    def is_implementation(self):
```
Comment (stonier) [inline rocon_app_utilities/src/rocon_app_utilities/rapp.py]: Again, direct variable access is ok.

File: rocon_app_utilities/src/rocon_app_utilities/rapp_cmd.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    print("Childs")
+    pass
+
+def _rapp_cmd_profile(argv):
+    indexer = RappIndexer()
+    pass
+
+def _rapp_cmd_compat(argv):
+
+    #  Parse command arguemtns
```
Comment (stonier) [inline rocon_app_utilities/src/rocon_app_utilities/rapp_cmd.py]: arguemtns -> arguments

**Comment by stonier** [comment]
Some comments above. Can you rebase everything into a single commit too?

Nice job.

File: rocon_app_utilities/src/rocon_app_utilities/rapp.py (hunk 5 above)
Comment (jihoonl) [inline rocon_app_utilities/src/rocon_app_utilities/rapp.py]: This will be renamed somehow.

File: rocon_app_utilities/src/rocon_app_utilities/rapp.py (hunk 7 above)
Comment (jihoonl) [inline rocon_app_utilities/src/rocon_app_utilities/rapp.py:182]: No. It is called from indexer not from internal function. Thus, it should be external method.

File: rocon_app_utilities/src/rocon_app_utilities/rapp.py (hunk 5 above)
Comment (jihoonl) [inline rocon_app_utilities/src/rocon_app_utilities/rapp.py]: Not rapp includes three name fields.
- `resource_name` 
- `parent_name`
- `ancestor_name`

**Comment by jihoonl** [comment]
I got 3 commits. I am scared of touching more since I have merged Marcus's code already here..

**Comment by stonier** [comment]
Most of marcus' stuff is in, so you can keep going here. He's just got small items to attack which he will add later.

File: rocon_app_manager/src/rocon_app_manager/rapp_manager.py [hunk 10]
```diff
@@ -70,13 +70,10 @@ def __init__(self):
...
         self._initialising_services = False
-
-        self._preinstalled_apps = self._load_installed_rapps()  # parses exported rapps from the package path
-        (platform_filtered_apps, capabilities_filtered_apps) = self._determine_runnable_rapps(self._preinstalled_apps)
+        self._indexer = rocon_app_utilities.RappIndexer(package_whitelist=self._param['rapp_package_whitelist'], package_blacklist=self._param['rapp_package_blacklist'])
+        self._runnable_apps, self._platform_filtered_apps, self._capabilities_filtered_apps = self._determine_runnable_rapps()
         self._init_services()
         self._private_publishers = self._init_private_publishers()
-        self._publish_app_list()
-        self._publishers['incompatible_app_list'].publish([], [], self._get_app_msg_list(platform_filtered_apps), self._get_app_msg_list(capabilities_filtered_apps))
```
Comment (stonier) [inline rocon_app_manager/src/rocon_app_manager/rapp_manager.py:79]: The incompatible app list was scheduled to publish only at init time - it is not necessary elsewhere. I notice this PR changes this behaviour - is it now required to update this publisher repeatedly?

File: rocon_app_manager/src/rocon_app_manager/rapp_manager.py [hunk 11]
```diff
@@ -611,7 +608,7 @@ def _monitor_rapp(self):
 
     def _load(self, directory, typ):
```
Comment (stonier) [inline rocon_app_manager/src/rocon_app_manager/rapp_manager.py]: I don't think this function is used anymore - perhaps it can be deleted.

File: rocon_app_manager/src/rocon_app_manager/rapp_manager.py [hunk 12]
```diff
@@ -674,6 +671,14 @@ def _flip_connections(self, remote_name, connection_names, connection_type, canc
             else:
                 rospy.logerr("App Manager : failed to flip [%s, %s]" % (resp.result, resp.error_message))
 
+    def _publish_current_rapp_status(self):
+        # Publish currently available rapp list
+        self._publish_app_list()
+
+        # Publish unavailable rapp list
+        # TODO: Currently blacklisted apps and non-whitelisted apps are not provided yet
+        self._publishers['incompatible_app_list'].publish([], [], self._get_app_msg_list(self._platform_filtered_apps), self._get_app_msg_list(self._capabilities_filtered_apps))
```
Comment (stonier) [inline rocon_app_manager/src/rocon_app_manager/rapp_manager.py]: Again, don't think we need to repetitively publish this one.

File: rocon_app_manager/src/rocon_app_manager/rapp_manager.py (hunk 10 above)
Comment (jihoonl) [inline rocon_app_manager/src/rocon_app_manager/rapp_manager.py:79]: No still it should publish once only in initialization. Forgot to add a method in.

**Comment by stonier** [comment]
@jihoonl why did you close this? I was about to merge and test.
